### PR TITLE
bpo-46170: Improve the error message when subclassing NewType

### PR DESCRIPTION
**Pull request by Gobot1234**
https://bugs.python.org/issue46170

### CODE REVIEW

**Comment by AlexWaygood** [comment]
You can use https://blurb-it.herokuapp.com/ to add a News entry 🙂

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -4005,6 +4005,15 @@ def test_missing__name__(self):
                 )
         exec(code, {})
 
+    def test_subclassing(self):
+        with self.assertRaisesRegex(
+            TypeError, 
+            f"You cannot subclass an instance of NewType, perhaps you were looking for:\n"
```
Comment (sobolevn) [inline Lib/test/test_typing.py]: Let's use the same message we use for other types: https://github.com/python/cpython/blob/d0b0ec758cba80c38cdb78126d04314fdbe76c40/Lib/typing.py#L1100

Maybe with an extra note about `ProUserId = NewType('ProUserId', UserId)`

Comment (sobolevn) [inline Lib/test/test_typing.py]: It feels like `re.escape()` might be a good fit here.

Comment (Gobot1234) [reply Lib/test/test_typing.py]:
```suggestion
            f"Cannot subclass {self!r}, perhaps you were looking for:\n"
```
Would this be better?

Comment (sobolevn) [reply Lib/test/test_typing.py]: I like it 👍

File: Lib/typing.py [hunk 2]
```diff
@@ -2459,7 +2459,7 @@ def __init__(self, name, tp):
 
     def __mro_entries__(cls, bases):
         raise TypeError(
-            f'You cannot subclass an instance of NewType, perhaps you were looking for:\n'
+            f'Cannot subclass {cls.__name__}, perhaps you were looking for:\n'
```
Comment (sobolevn) [inline Lib/typing.py]: I also vote for a single-line version of this message, no need to use `\n`. Do you agree?

Comment (Gobot1234) [reply Lib/typing.py]: I think it's clear and easy to copy and paste if it's two lines

**Comment by Gobot1234** [comment]
I've just realised that what I thought was the `cls` param to `__mro_entries__` isn't the class that's subclassing it's the instance, totally my bad. I'm not entirely sure how to get the name of the subclassing type without some byte-code hack. Any advice?

**Comment by Gobot1234** [comment]
I've had the idea to return a dummy class from mro entries and have the dummy class implement init subclass which then raises the type error.

**Comment by AlexWaygood** [comment]
> I've had the idea to return a dummy class from mro entries and have the dummy class implement init subclass which then raises the type error.

That looks like it might work, but I'd beware of making the implementation too complex for `NewType`. The whole point of `NewType` is that it's meant to be a super-simple thing that does basically nothing at runtime. A better error message would be a great thing to have, but I doubt it would be accepted if it comes at the cost of a much more complex implementation (and I think probably has a very low chance of being accepted if it has a performance penalty).

**Comment by gvanrossum** [comment]
What Alex says.

**Comment by Gobot1234** [comment]
Should I update this to use `BaseException.__note__`?

**Comment by AlexWaygood** [comment]
> Should I update this to use `BaseException.__note__`?

Hmm, I don't think this is really the use case for which that was designed. How exactly are you thinking it might help? 🙂

File: Lib/typing.py [hunk 3]
```diff
@@ -2457,6 +2457,19 @@ def __init__(self, name, tp):
...
 
+    def __mro_entries__(cls, bases):
+        new_type_name = cls.__name__
+
+        class Dummy:
+            def __init_subclass__(cls):
+                current_name = cls.__name__
+                raise TypeError(
+                    f'Cannot subclass {new_type_name}, perhaps you were looking for: '
+                    f'{current_name} = NewType({current_name!r}, {new_type_name})'
```
Comment (AlexWaygood) [inline Lib/typing.py]: This error message tells me _what_ to do instead, but doesn't tell me _why_ the error is being raised (it's an instance of `NewType`).

File: Lib/typing.py [hunk 4]
```diff
@@ -2457,6 +2457,19 @@ def __init__(self, name, tp):
         if def_mod != 'typing':
             self.__module__ = def_mod
 
+    def __mro_entries__(cls, bases):
+        new_type_name = cls.__name__
```
Comment (AlexWaygood) [inline Lib/typing.py]: I find this variable name really confusing. You've called it `new_type_name`, but it's actually the name of the _old_ class -- the class that's being subclassed.

**Comment by Gobot1234** [comment]
I'm personally a bit fan of being told how to fix something but if you think the implementation is too much added maintenance I'd be happy to remove it

File: Lib/typing.py (hunk 4 above)
Comment (Gobot1234) [reply Lib/typing.py]: I thought new_type_name was alright cause its the name of the `NewType` but I'm open to suggestions

**Comment by Gobot1234** [comment]
> > Should I update this to use `BaseException.__note__`?
> 
> Hmm, I don't think this is really the use case for which that was designed. How exactly are you thinking it might help? 🙂

I thought the point of the added attribute was to enrich exceptions.

File: Lib/typing.py (hunk 4 above)
Comment (AlexWaygood) [reply Lib/typing.py]: Yeah, I get where you got the variable name from! But I kept on tripping up on it when I was playing around with the patch. Why not use `superclass_name`/`subclass_name`? :)

File: Lib/typing.py (hunk 3 above)
Comment (Gobot1234) [inline Lib/typing.py:2468]:
```suggestion
                    f'Cannot subclass an instance of NewType, perhaps you were looking for: '
                    f'{current_name} = NewType({current_name!r}, {new_type_name})'
```

Comment (Gobot1234) [reply Lib/typing.py]:
```suggestion
                    f'Cannot subclass an instance of NewType, perhaps you were looking for: '
                    f'{current_name} = NewType({current_name!r}, {new_type_name})'
```

Comment (AlexWaygood) [reply Lib/typing.py]: That looks a lot better to me!

**Comment by AlexWaygood** [comment]
> > > Should I update this to use `BaseException.__note__`?
> > 
> > 
> > Hmm, I don't think this is really the use case for which that was designed. How exactly are you thinking it might help? 🙂
> 
> I thought the point of the added attribute was to enrich exceptions.

It is, of course, but the intended use case (as far as I understand PEP 678) is generally to add information to exceptions that "make sense". The error message when you try to subclass an instance of `NewType` is currently a bit nonsensical: it tells you that you've supplied the wrong number of arguments to `NewType.__init__`. I think we want to replace that error message entirely, rather than add additional information to it.

File: Lib/test/test_typing.py [hunk 5]
```diff
@@ -4005,6 +4005,17 @@ def test_missing__name__(self):
                 )
         exec(code, {})
 
+    def test_subclassing(self):
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
    def test_error_message_when_subclassing(self):
```

File: Lib/test/test_typing.py [hunk 6]
```diff
@@ -4005,6 +4005,17 @@ def test_missing__name__(self):
...
+        with self.assertRaisesRegex(
+            TypeError,
+            re.escape(
+                f"Cannot subclass an instance of NewType, perhaps you were looking for: "
+                f"ProUserId = NewType('ProUserId', UserId)"
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
                f"Cannot subclass an instance of NewType. Perhaps you were looking for: "
                f"`ProUserId = NewType('ProUserId', UserId)`"
```

File: Lib/typing.py [hunk 7]
```diff
@@ -2457,6 +2457,19 @@ def __init__(self, name, tp):
...
 
+    def __mro_entries__(cls, bases):
+        superclass_name = cls.__name__
+
+        class Dummy:
+            def __init_subclass__(cls):
+                subclass_name = cls.__name__
+                raise TypeError(
+                    f"Cannot subclass an instace of NewType, perhaps you were looking for: "
+                    f"{subclass_name} = NewType({subclass_name!r}, {superclass_name})"
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
                    f"Cannot subclass an instace of NewType. Perhaps you were looking for: "
                    f"`{subclass_name} = NewType({subclass_name!r}, {superclass_name})`"
```

File: Lib/typing.py [hunk 8]
```diff
@@ -2457,6 +2457,19 @@ def __init__(self, name, tp):
         if def_mod != 'typing':
             self.__module__ = def_mod
 
+    def __mro_entries__(cls, bases):
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
    def __mro_entries__(cls, bases):
    # We define __mro_entries__ to get a better error message
    # if a user attempts to subclass a NewType instance. bpo-46170
```

**Review by AlexWaygood — COMMENTED** [review]
Personally, I would still go for the simpler implementation myself, but this is looking a lot better to me now 👍

File: Lib/typing.py [hunk 9]
```diff
@@ -2458,14 +2458,14 @@ def __init__(self, name, tp):
...
+        superclass_name = cls.__name__
 
         class Dummy:
             def __init_subclass__(cls):
-                current_name = cls.__name__
+                subclass_name = cls.__name__
                 raise TypeError(
-                    f'Cannot subclass {new_type_name}, perhaps you were looking for: '
-                    f'{current_name} = NewType({current_name!r}, {new_type_name})'
+                    f"Cannot subclass an instace of NewType, perhaps you were looking for: "
```
Comment (arhadthedev) [inline Lib/typing.py]:
```suggestion
                    f"Cannot subclass an instance of NewType, perhaps you were looking for: "
```

Comment (Gobot1234) [reply Lib/typing.py]: I thought I fixed this, thanks

Comment (Gobot1234) [reply Lib/typing.py]: Oh what it's already been fixed

Comment (AlexWaygood) [reply Lib/typing.py]: @arhadthedev I think you're looking at the changes as they were a few commits back

**Review by JelleZijlstra — APPROVED** [review]
This is a creative abuse of `__mro_entries__`, but it doesn't seem likely to cause maintenance or performance issues and it gives a much nicer error message. I'm fine with moving forward with it.

File: Misc/NEWS.d/next/Library/2021-12-26-14-45-51.bpo-46170.AQ7kSM.rst [hunk 10]
```diff
@@ -0,0 +1 @@
+Improve the error message when you try to subclass an instance of typing.NewType
```
Comment (sobolevn) [inline Misc/NEWS.d/next/Library/2021-12-26-14-45-51.bpo-46170.AQ7kSM.rst]:
```suggestion
Improve the error message when you try to subclass an instance of :class:`typing.NewType`
```

File: Misc/NEWS.d/next/Library/2021-12-26-14-45-51.bpo-46170.AQ7kSM.rst [hunk 11]
```diff
@@ -0,0 +1 @@
+Improve the error message when you try to subclass an instance of :class:`typing.NewType`
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2021-12-26-14-45-51.bpo-46170.AQ7kSM.rst]:
```suggestion
Improve the error message when you try to subclass an instance of :class:`typing.NewType`.
```

**Review by Fidget-Spinner — APPROVED** [review]
This is such a clever workaround. I'm at a loss for words. My only minor concern is that it feels slightly hacky.

**Comment by Fidget-Spinner** [comment]
@JelleZijlstra I plan to merge, is that alright? Does anyone else have any objections?

**Comment by JelleZijlstra** [comment]
Sounds good, I was planning to hold off until the next alpha is out since it's bumpy and I don't want to give Pablo more headaches.

**Comment by Fidget-Spinner** [comment]
> Sounds good, I was planning to hold off until the next alpha is out since it's bumpy and I don't want to give Pablo more headaches.

Good idea, I forgot about that. Let's do that then :).

**Pull request merged**